core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala

```diff
@@ -917,4 +917,111 @@ class TaskSchedulerImplSuite extends SparkFunSuite with LocalSparkContext with B
       taskScheduler.initialize(new FakeSchedulerBackend)
     }
   }
+
+  test("Completions in zombie tasksets update status of non-zombie taskset") {
+    val taskScheduler = setupSchedulerWithMockTaskSetBlacklist()
+    val valueSer = SparkEnv.get.serializer.newInstance()
+
+    def completeTaskSuccessfully(tsm: TaskSetManager, partition: Int): Unit = {
+      val indexInTsm = tsm.partitionToIndex(partition)
+      val matchingTaskInfo = tsm.taskAttempts.flatten.filter(_.index == indexInTsm).head
+      val result = new DirectTaskResult[Int](valueSer.serialize(1), Seq())
+      tsm.handleSuccessfulTask(matchingTaskInfo.taskId, result)
+    }
+
+    // Submit a task set, have it fail with a fetch failed, and then re-submit the task attempt,
+    // two times, so we have three active task sets for one stage.  (For this to really happen,
+    // you'd need the previous stage to also get restarted, and then succeed, in between each
+    // attempt, but that happens outside what we're mocking here.)
+    val zombieAttempts = (0 until 2).map { stageAttempt =>
+      val attempt = FakeTask.createTaskSet(10, stageAttemptId = stageAttempt)
+      taskScheduler.submitTasks(attempt)
+      val tsm = taskScheduler.taskSetManagerForAttempt(0, stageAttempt).get
+      val offers = (0 until 10).map{ idx => WorkerOffer(s"exec-$idx", s"host-$idx", 1) }
+      taskScheduler.resourceOffers(offers)
+      assert(tsm.runningTasks === 10)
+      if (stageAttempt < 2) {
+        // fail attempt
+        tsm.handleFailedTask(tsm.taskAttempts.head.head.taskId, TaskState.FAILED,
+          FetchFailed(null, 0, 0, 0, "fetch failed"))
+        // the attempt is a zombie, but the tasks are still running (this could be true even if
+        // we actively killed those tasks, as killing is best-effort)
+        assert(tsm.isZombie)
+        assert(tsm.runningTasks === 9)
+      }
+      tsm
+    }
+
+    // we've now got 2 zombie attempts, each with 9 tasks still active.  Submit the 3rd attempt for
+    // the stage, but this time with insufficient resources so not all tasks are active.
+
+    val finalAttempt = FakeTask.createTaskSet(10, stageAttemptId = 2)
+    taskScheduler.submitTasks(finalAttempt)
+    val finalTsm = taskScheduler.taskSetManagerForAttempt(0, 2).get
+    val offers = (0 until 5).map{ idx => WorkerOffer(s"exec-$idx", s"host-$idx", 1) }
+    val finalAttemptLaunchedPartitions = taskScheduler.resourceOffers(offers).flatten.map { task =>
```

Comment:
Yet, launched tasks has nothing to do with other running tasks in other `TaskSet`s. But, is it possible to take those running tasks into consideration when launch a new task (in source code) ? For example,  launching FetchFailed task or tasks who do not have a running copy across `TaskSet`s firstly ?

(But, it seems we will always have running copies in other `TaskSet`s for our  final `TaskSet`, except FetchFailed task, right? It's more like we are not talking about resubmitting a stage, but resubmitting tasks who do not have running copies across previous `TaskSet`s.)

Reply:
we've previously debated about what to do with the tasks still running in a zombie attempt, and there hasn't been any definitive conclusion.  I'm just trying to do a correctness fix here.  Briefly, in general there is an expectation that those tasks are unlikely to succeed (because they won't be able to get their shuffle input, same as the original fetch failure), so we don't want to delay starting a new attempt of that task.  And perhaps we should even actively kill those tasks (you'll see comments about that in various places).  But if they do succeed, we need to handle them correctly.  Note that even if we did try to actively kill them, you'd still need to handle a late-completion, as killing would only be "best-effort".

Reply:
> because they won't be able to get their shuffle input, same as the original fetch failure

why? In `DAGScheduler`, we only unregister one MapStatus of parent stage, so other running tasks within the failed (child) stage (caused by a fetch fail task)  may still get MapOutputs from `MapOutputTrackerMaster`, and fetch data from other `Executor`s. So, they can success normally. 
Do I miss something?

Reply:
the assumption is that a fetchfailure means that *all* data on that host is unavailable.  As shuffles are all-to-all, its very likely that every task is going to need some piece of data from that host.  Its possible that they already grabbed all the data they need, before the problem occurred with the host, we don't know.  Also, there is no "partial progress" for a task -- tasks don't know how to grab all the shuffle output they can, then just wait until the missing bit becomes available again.  They fail as soon as the data they need is unavailable (with some retries, but there is no "pause" nor a check for data on another source). 

Also the dagscheduler is a little confusing on this -- it does the unregister in two parts (I have no idea why anymore, to be honest):

https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala#L1391

https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala#L1406

Reply:
The explanation is quite clear and I get understand now. Thank you very mush! @squito